pkg/reconcilermanager/controllers/reposync_controller.go

```diff
@@ -486,7 +509,34 @@ func (r *RepoSyncReconciler) SetupWithManager(mgr controllerruntime.Manager, wat
 			handler.EnqueueRequestsFromMapFunc(r.mapMembershipToRepoSyncs()),
 			builder.WithPredicates(predicate.GenerationChangedPredicate{}))
 	}
-	return controllerBuilder.Complete(r)
+
+	ctrlr, err := controllerBuilder.Build(r)
+	r.controller = &ctrlr
+
+	return err
+}
+
+func (r *RepoSyncReconciler) watchConfigMaps(rs *v1beta1.RepoSync) error {
+	// We add watches dynamically at runtime based on the RepoSync namespace
+	// in order to avoid watching ConfigMaps in the entire cluster.
+	if rs.Spec.SourceType != string(v1beta1.HelmSource) || rs.Spec.Helm == nil ||
```

Comment:
Do we remove watches when it's no longer needed (e.g. when the RSync in that namespace is removed or updated to not use ConfigMaps)?

Reply:
This PR does not do that, but I can add that. I think we'll have to keep a map of RSync to watched namespaces to keep track of that. I agree that removing unneeded watches is for the best, so I will make the change.

Reply:
I looked into it a bit more and it seems that there is no current way to stop a watch once you've started one. There's an open PR and an open issue in controller-runtime, with updates as recent as 3 days ago:

https://github.com/kubernetes-sigs/controller-runtime/pull/2159
https://github.com/kubernetes-sigs/controller-runtime/issues/1884

I added a TODO to this function instead to remind is to keep an eye on that. I found some other similar issues (e.g. https://github.com/opensergo/opensergo-control-plane/issues/23) who are also waiting for the fix.

Reply:
As discussed offline, I looked at [the custom watcher in Config Sync](https://github.com/GoogleContainerTools/kpt-config-sync/blob/21100ec59b8419fbe8fbd05b21933c72aebf6d8e/pkg/remediator/watch/manager.go#L79), but it looks pretty specific to the reconciler so I don't think it supports this use case. I think the best option is to leave the watches for now and wait for the fix in controller-runtime for the future.

Reply:
Sam, what I'm trying to say is that the we're merging vs. overriding values of the same keys in different values files, so the mode types makes sense to me.

Reply:
FYI I came across this issue that talks about a feature request for merge instead of override https://github.com/helm/helm/issues/3486

Reply:
Maybe the merge strategies could be named `shallow`/`deep`?

Reply:
As discussed offline, we will live it as `override` and `merge`, but rename the field to `valuesKeyMergeMode`.